### PR TITLE
GH-34610: [Java] Fix valueCount when loading NullVector

### LINKED ISSUE #34610: [Java] When loading NullVector, valueCount should be properly set
### Describe the bug, including details regarding any error messages, version, and platform.

In `NullVector.loadFieldBuffers` we currently don't set `valueCount`, which is incorrect.

### Component(s)

Java

### PR DESCRIPTION
**Pull request by sunchao**
We should set `valueCount` and `nullCount` when loading `NullVector`.

### Rationale for this change

When importing a `NullVector`, for example, via Arrow C interface, the `valueCount` and `nullCount` is currently not set properly, which will result both values to be 0.

### What changes are included in this PR?

Set `valueCount` when loading the vector.

### Are these changes tested?

Added a test. Also had to add extra constructor to facilitate the test.

### Are there any user-facing changes?

No.

* Closes: #34610

### CODE REVIEW

**Review by lidavidm — APPROVED** [review]

**Review by viirya — APPROVED** [review]

**Comment by viirya** [comment]
```
Error:  Errors: 
Error:    TestBasicOperation.getStreamLargeBatch:273->test:315->test:333 � IllegalState Memory was leaked by query. Memory leaked: (134217728)
Allocator(ROOT) 0/134217728/270532608/9223372036854775807 (res/actual/peak/limit)
```

Hmm, is this related?

**Comment by lidavidm** [comment]
No, there is a long standing issue for it that nobody has had time to look at.

**Comment by lidavidm** [comment]
The JNI pipeline failure does seem related, though

**Comment by viirya** [comment]
yea, that's the added test `testNullVector`.

```
Error:  org.apache.arrow.c.RoundtripTest.testNullVector  Time elapsed: 0.008 s  <<< FAILURE!
org.opentest4j.AssertionFailedError: expected: <true> but was: <false>
```

**Comment by amol-** [comment]
Closing because it has been untouched for a while, in case it's still relevant feel free to reopen and move it forward 👍

File: java/c/src/test/java/org/apache/arrow/c/RoundtripTest.java [hunk 1]
```diff
@@ -462,6 +462,13 @@ public void testUInt8Vector() {
     }
   }
 
+  @Test
+  public void testNullVector() {
+    try (final NullVector vector = new NullVector("v", 1024)) {
+      assertTrue(roundtrip(vector, NullVector.class));
```
Comment (danepitkin) [inline java/c/src/test/java/org/apache/arrow/c/RoundtripTest.java:468]: Hmm, its not obvious to me why this test is failing. I would've expected your change to `loadFieldBuffers()` to work. Maybe the field comparison is failing? I think you could try passing a custom `TypeEqualsVisitor` with `checkName` or `checkMetadata` disabled to narrow things down on line 155.

```
return VectorEqualsVisitor.vectorEquals(vector, imported, <custom_visitor_here>);
```

**Comment by sunchao** [comment]
Sorry @amol- and @danepitkin , we found a workaround to this issue internally so I haven't got a chance to re-visit the test failure. It's indeed weird and I plan to reproduce it locally and see what happened.

**Comment by danepitkin** [comment]
> Sorry @amol- and @danepitkin , we found a workaround to this issue internally so I haven't got a chance to re-visit the test failure. It's indeed weird and I plan to reproduce it locally and see what happened.

No worries! I noticed the Java implementation has been a bit quiet in general, so its great to see new fixes coming in.

**Comment by danepitkin** [comment]
I've followed this PR up here: https://github.com/apache/arrow/pull/38973

The issue was in NullVector.getTransferPair(). It was not passing the name when transferring, which resulted in this check failing when comparing field names: https://github.com/apache/arrow/blob/61a4a803779db17a6c7bd57210e0dc92a067f2a0/java/vector/src/main/java/org/apache/arrow/vector/compare/TypeEqualsVisitor.java#L134

I created a new PR since I didn't have permissions to push to your branch. I hope you don't mind!

**Pull request closed without merging**